lib/active_storage_validations/marcel_extensor.rb

```diff
@@ -1,5 +1,7 @@
 # frozen_string_literal: true
 
+require "marcel"
```

Comment:
`unless defined?(Marcel)`

Reply:
Hey @Mth0158 — this isn’t a normal pattern in Ruby. If you call require on the same path that has already been loaded, Ruby will no-op very quickly. Probably more quickly that checking if a constant is defined since it’s a hash lookup vs a constant lookup.

Reply:
That said, this is your project and if you think that’s better, I’m happy to do it. If you look in rails or other prominent Ruby code based, I’d be surprised if you see anyone do this.

Reply:
@ngan you are absolutely right, I was not aware of the no-op mecanism (https://ruby-doc.org/3.3.6/Kernel.html#method-i-require).

@igorkasyanchuk that what we were discussing in the other PR, might be a good idea to remove the `unless defined?(MiniMagick/Vips)` in our code?

Reply:
Not a blocker from my side, but defined works faster. Or I'm missing something?

```ruby
require "benchmark"

def require_check
  require "marcel"
end

def defined_check
  defined?(Marcel)
end

Benchmark.bm do |x|
  x.report("require") { 1000.times { require_check } }
  x.report("defined") { 1000.times { defined_check } }
end

```

```
ruby app.rb
             user     system      total        real
require  0.025928   0.019764   0.045692 (  0.045691)
defined  0.000067   0.000000   0.000067 (  0.000068)
```

Reply:
Mmm I would rather do the below, but the `defined_check` is still better

```ruby
require "benchmark"

require "marcel"

def require_check
  require "marcel"
end

def defined_check
  require "marcel" unless defined?(Marcel)
end

Benchmark.bm do |x|
  x.report("require") { 1000.times { require_check } }
  x.report("defined") { 1000.times { defined_check } }
end
```

```sh
             user     system      total        real
require  0.000462   0.000007   0.000469 (  0.000468)
defined  0.000042   0.000000   0.000042 (  0.000046)
```

Reply:
I haven't benchmarked it myself, but I think you'd want to bench mark with `marcel` already loaded such that all calls to `require` will no-op, and calls to `defined?(Marcel)` will return the constant, instead of `nil`. I also imagine it depends on the size of the app / running process since you'd have a lot more constants loaded. But again, I'm not sure. I've just never seen the practice in my 18 years of working with Ruby 😅. Maybe this is a good question for the Ruby community? https://www.ruby-lang.org/en/community/

Reply:
Ah yeah, looks like I was writing my comment on a stale page. But yeah, @Mth0158 is a better benchmark. I image `defined?(Marcel)` might get _slightly_ slower for real-world apps. It looks like we're now in micro-optimization territory. In which case, I would argue that our objective here is to "load the marcel library" and not "check if the constant Marcel exists." Since...`Marcel` is just a constant that could be defined by anything/anywhere...even a monkey-patch.

Anyways, this was such a fun conversation -- thanks guys! Now I'm really curious what Ruby-core would think/say.

Reply:
I am really curious about it 🤔. It is all about the perf of calling `require "marcel"` vs calling `defined?(Marcel)`. In the benchmark I have done we see that `defined?(Marcel)` is a 10x faster, but it now in the context of an app that will have many constants defined.

I guess that, making the assumption that `require` and `defined?` perform the check at the same speed, it would be faster to use `defined?` because they are probably less than 10 constant definitions per file (therefore per required file). But that's a lot hypothetical.

Probably that not seing this `defined?` pattern is a sign that's not the most optimised move?

Reply:
Not sure if you're in Nate's Rails Performance Slack, but I posted the question there: https://railsperf.slack.com/archives/C0QKHT5LP/p1734109484198909

Reply:
Thanks @ngan, curious to read the answer :).
I can't join the Slack because it's restricted to some organisations, or you have to be invited by an admin. I would gladly join the slack channel, I am really interesting in theses questions :)

<img width="720" alt="image" src="https://github.com/user-attachments/assets/da9172ac-107d-4fae-96ed-fad9a2f4b569" />

Reply:
I've asked the Admin to extend you an invite. See you on the other side.

Reply:
Until then, one member mentioned this:
https://github.com/igorkasyanchuk/active_storage_validations/blob/1b2d5dc1353276f0c4fa59f48d549aaf20a10b46/lib/active_storage_validations/analyzer/image_analyzer/image_magick.rb#L12-L17

This is somewhat dangerous because:
> It's possible someone might open up the MiniMagic namespace to do like MiniMagic::CustomThing without actually loading the library (it's possible to do that with Modules... not great, but possible)

Reply:
👋🏻 💖 The insight I want to share is that:

- `defined?` checks whether a constant exists
- `require` loads a file once, and tracks what has been loaded already via `$LOADED_FEATURES`.

There is some overlap between files and constants, but in most real-world applications, there is deviation too (sometimes reasonable, sometimes unreasonable). As a gem, I would recommend resiliency over correctness, and just use `require`.